matsim/src/main/java/org/matsim/analysis/ModeStatsControlerListener.java

```diff
@@ -136,31 +127,50 @@ private void collectModeShareInfo(final IterationEndsEvent event) {
 		for ( Double val : this.modeCnt.values() ) {
 			sum += val ;
 		}
-		
+
+		// add new modes not encountered in previous iterations
+		this.modes.addAll(modeCnt.keySet());
+
+		// calculate and save this iteration's mode shares
+		log.info("Mode shares over all " + sum + " trips found. MainModeIdentifier: " + mainModeIdentifier.getClass());
+		for ( String mode : modes ) {
+			Double cnt = this.modeCnt.getOrDefault(mode, 0.0) ;
+			double share = 0. ;
+			if ( cnt!=null ) {
+				share = cnt/sum;
+			}
+			log.info("-- mode share of mode " + mode + " = " + share );
+
+			Map<Integer, Double> modeHistory = this.modeHistories.get(mode) ;
+			if ( modeHistory == null ) {
+				modeHistory = new TreeMap<>() ;
+				for (int iter = firstIteration; iter < event.getIteration(); iter++) {
+					modeHistory.put(iter, 0.0);
+				}
+				this.modeHistories.put(mode, modeHistory) ;
+			}
+			modeHistory.put( event.getIteration(), share ) ;
+		}
+
+		BufferedWriter modeOut = IOUtils.getBufferedWriter(this.modeFileName + ".txt");
 		try {
-			this.modeOut.write( String.valueOf(event.getIteration()) ) ;
-			log.info("Mode shares over all " + sum + " trips found. MainModeIdentifier: " + mainModeIdentifier.getClass());
+			modeOut.write("Iteration");
 			for ( String mode : modes ) {
-				Double cnt = this.modeCnt.get(mode) ;
-				double share = 0. ;
-				if ( cnt!=null ) {
-					share = cnt/sum;
-				}
-				log.info("-- mode share of mode " + mode + " = " + share );
-				this.modeOut.write( "\t" + share ) ;
-				
-				Map<Integer, Double> modeHistory = this.modeHistories.get(mode) ;
-				if ( modeHistory == null ) {
-					modeHistory = new TreeMap<>() ;
-					this.modeHistories.put(mode, modeHistory) ;
+				modeOut.write("\t" + mode);
+			}
+			modeOut.write("\n"); ;
+			for (int iter = firstIteration; iter <= event.getIteration(); iter++) {
+				modeOut.write( String.valueOf(iter) ) ;
+				for ( String mode : modes ) {
+					modeOut.write( "\t" + modeHistories.get(mode).get(iter)) ;
```

Comment:
since we are touching it :-)
Perhaps we could round this double to a more readable form, e.g. 4 digits?

Reply:
I'm not so sure about this. We have cases which need more digits, e.g. you have your entire Switzerland model and enable drt in a small area only. With 4 digits you end up with a mode share of 00.00 for drt and don't know whether it was used at all whereas with the non-rounded value you can multiply with the share of trips in that small area of all trips in Switzerland and have a useful mode share for drt.

Reply:
Mhm :-) I would argue that anything after two-digits behind the percentage-comma is either noise or requires a dedicated analysis anyway

Reply:
Yes, sure, in many cases it is rather noise, however in that rarely used mode problem it is not. We have this a lot with drt (mode share somewhere in the 10^⁻4 or 10^-5 area). Some information on intermediate iterations is often helpful. Unless you write out plan files in each iteration it is impossible to get mode shares for intermediate iterations and get some idea whether agents even tried out drt or never tried because something was broken or mode shares have not converged yet.

Reply:
So maybe this should be made configurable? Maybe in a separate pull request (making it configurable for all output files)?